File: backend/src/models/Dependent.js
const { DataTypes } = require('sequelize');

// Define allowed relationship values
const RELATIONSHIP_VALUES = {
  SON: 'Son',
  DAUGHTER: 'Daughter', 
  SPOUSE: 'Spouse',
  PARENT: 'Parent',
  SIBLING: 'Sibling',
  OTHER: 'Other'
};

module.exports = (sequelize) => {
  const Dependent = sequelize.define('Dependent', {
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true
    },
    memberId: {
      type: DataTypes.BIGINT,
      allowNull: false,
      references: {
        model: 'members',
        key: 'id'
      }
    },
    linkedMemberId: {
      type: DataTypes.BIGINT,
      allowNull: true,
      references: {
        model: 'members',
        key: 'id'
      }
    },
    firstName: {
      type: DataTypes.STRING(100),
      allowNull: false
    },
    middleName: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    lastName: {
      type: DataTypes.STRING(100),
      allowNull: false
    },
    dateOfBirth: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    gender: {
      type: DataTypes.STRING(10),
      allowNull: true
    },
    relationship: {
      type: DataTypes.ENUM(Object.values(RELATIONSHIP_VALUES)),
      allowNull: true,
      validate: {
        isIn: {
          args: [Object.values(RELATIONSHIP_VALUES)],
          msg: 'Relationship must be one of: Son, Daughter, Spouse, Parent, Sibling, Other'
        }
      }
    },
    phone: {
      type: DataTypes.STRING(20),
      allowNull: true
    },
    email: {
      type: DataTypes.STRING(255),
      allowNull: true,
      validate: {
        isEmail: true
      }
    },
    baptismName: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    isBaptized: {
      type: DataTypes.BOOLEAN,
      defaultValue: false
    },
    baptismDate: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    nameDay: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    medicalConditions: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    allergies: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    medications: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    dietaryRestrictions: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    languagePreference: {
      type: DataTypes.STRING(50),
      allowNull: true
    },
    interestedInServing: {
      type: DataTypes.STRING(10),
      allowNull: false,
      defaultValue: 'no'
    },
    notes: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    createdAt: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    },
    updatedAt: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    }
  }, {
    tableName: 'dependents',
    timestamps: true,
    underscored: true
  });

  Dependent.associate = (models) => {
    Dependent.belongsTo(models.Member, {
      foreignKey: 'memberId',
      as: 'member'
    });

    // When a dependent self-claims, it links to a Member account
    Dependent.belongsTo(models.Member, {
      foreignKey: 'linkedMemberId',
      as: 'linkedMember'
    });
  };

  // Export the relationship values for use in other parts of the application
  Dependent.RELATIONSHIP_VALUES = RELATIONSHIP_VALUES;

  return Dependent;
}; 

File: backendJava/src/main/java/church/abunearegawi/backend/model/Dependent.java
package church.abunearegawi.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "dependents")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Dependent {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "linked_member_id")
    private Member linkedMember;

    @Column(name = "first_name", nullable = false, length = 100)
    private String firstName;

    @Column(name = "middle_name", length = 100)
    private String middleName;

    @Column(name = "last_name", nullable = false, length = 100)
    private String lastName;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Column(length = 10)
    private String gender;

    @Enumerated(EnumType.STRING)
    private Relationship relationship;

    @Column(length = 20)
    private String phone;

    @Column(length = 255)
    private String email;

    @Column(name = "baptism_name", length = 100)
    private String baptismName;

    @Column(name = "is_baptized")
    @Builder.Default
    private boolean isBaptized = false;

    @Column(name = "baptism_date")
    private LocalDate baptismDate;

    @Column(name = "medical_conditions", columnDefinition = "TEXT")
    private String medicalConditions;

    @Column(columnDefinition = "TEXT")
    private String allergies;

    @Column(columnDefinition = "TEXT")
    private String medications;

    @Column(name = "dietary_restrictions", columnDefinition = "TEXT")
    private String dietaryRestrictions;

    @Column(name = "language_preference", length = 50)
    private String languagePreference;

    @Column(name = "interested_in_serving", length = 10, nullable = false)
    @Builder.Default
    private String interestedInServing = "no";

    @Column(columnDefinition = "TEXT")
    private String notes;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum Relationship {
        Son, Daughter, Spouse, Parent, Sibling, Other
    }
}
